zetta_utils/mazepa_layer_processing/common/blendable_apply_flow.py

```diff
@@ -0,0 +1,158 @@
+@mazepa.taskable_operation
+def reduce_by_sum(
+    src_layers: List[VolumetricLayer], idx: VolumetricIndex, dst: VolumetricLayer
+) -> None:
+    assert len(src_layers) > 0
+    res = src_layers[0][idx]
+    for layer in src_layers[1:]:
+        res = tensor_ops.common.add(res, layer[idx])
+    dst[idx] = res
+
+
+@builder.register(
+    "BlendableApplyFlowSchema",
+    cast_to_vec3d=["dst_resolution"],
+    cast_to_intvec3d=["chunk_size", "blend_pad"],
+)
+@mazepa.flow_schema_cls
+@attrs.mutable
+class BlendableApplyFlowSchema(Generic[P, R_co]):
+    operation: BlendableOpProtocol[P, R_co]
+    chunk_size: IntVec3D
+    dst_resolution: Vec3D
+    blend_pad: Optional[IntVec3D] = None
+    blend_mode: Literal["linear", "quadratic"] = "linear"
+    temp_layers_dir: Optional[str] = None
+    chunker: VolumetricIndexChunker = attrs.field(init=False)
+
+    def __attrs_post_init__(self):
+        self.chunker = VolumetricIndexChunker(
+            chunk_size=self.chunk_size, resolution=self.dst_resolution
+        )
+
+    #    chunker in post init
+    def flow(  # pylint: disable=too-many-locals
+        self,
+        idx: VolumetricIndex,
+        dst: VolumetricLayer,
+        *args: P.args,
+        **kwargs: P.kwargs,
+    ) -> mazepa.FlowFnReturnType:
+        # can't figure out how to force mypy to check this
+        assert len(args) == 0
+        logger.info(f"Breaking {idx} into chunks with {self.chunker}.")
+        idx_chunks = self.chunker(idx)
+        # case without blending
+        if self.blend_pad is None:
+            tasks = [
+                self.operation.make_task(
+                    idx=idx_chunk,
+                    **kwargs,  # type: ignore
+                )
+                for idx_chunk in idx_chunks
+            ]
+            logger.info(
+                f"Submitting {len(tasks)} processing tasks from operation {self.operation}."
+            )
+            yield tasks
+        # case with blending
+        else:
+            if not isinstance(self.temp_layers_dir, str):
+                raise ValueError("`temp_layers_dir` must be specified when using blending")
+            # padding for the edges of the idx so they don't get affected by blending
+            idx_padded = idx.pad(self.blend_pad)
+            logger.info("Writing to temporary destinations.")
+            blending_weighter = VolumetricDataBlendingWeighter(self.blend_pad, self.blend_mode)
+
+            tasks = []
+            idx_chunk_temps: List[List[VolumetricLayer]] = [[] for _ in idx_chunks]
+            for chunker, chunker_idx in self.chunker.split_into_nonoverlapping_chunkers(
+                self.blend_pad
+            ):
+                dst_temp = dst.clone(
+                    name=path.join(
+                        self.temp_layers_dir,
+                        f"_{self.operation.__class__.__name__}_temp_{chunker_idx}",
+                    )
+                )
+                dst_temp.write_preprocs = dst.write_preprocs.copy()
```

Comment:
Does `clone` copy lists by reference? This could be dangerous, as I wouldn't have thought to copy the preproc list here. Can we somehow avoid this and/or make it more obvious that this is the behavior? At the very least, we could put it in the comment for `layer.clone`.

Reply:
`attrs.evolve` does indeed copy lists by reference. This caught me off-guard as well. I could add deepcopy to `layer.clone`, or I could just add copy to the lists in `layer` (and add warning comments in `layer.clone` and `backend.clone`). I think the latter is better since if I added a deepcopy to `layer.clone` then `backend.clone` should probably also have deepcopy and that's waste of copying.